### ISSUE TITLE
Make Chronicle-Wire backwards-compatible in format

### ISSUE THREAD

**Issue by leventov** [OpenHFT/Chronicle-Wire#25]
Somewhere between Wire 1.3.6 and present versions compatibility was broken, so e. g. JSON produced by Wire 1.3.6 is not readable by newer versions. This makes impossible to use Chronicle Map 3 together with present (if any) versions of Chronicle Queue: http://stackoverflow.com/questions/37954548/exception-while-reading-from-the-chronicle-map-file#comment63402779_37957959

**Comment by peter-lawrey** [comment]
Do you have an example of the JSON it can't handle? I could understand a
new version not being able to read by an older version but a newer version
not handling an older one is surprising.  Could you add an error message
which dumps the entire JSON if an exception is thrown?

**Comment by leventov** [comment]
@peter-lawrey here is a failing test:
```java
package net.openhft.chronicle.wire;

import net.openhft.chronicle.bytes.Bytes;
import net.openhft.chronicle.core.io.IORuntimeException;
import net.openhft.chronicle.core.util.ReadResolvable;
import org.jetbrains.annotations.NotNull;
import org.junit.Assert;
import org.junit.Test;

public class EnumTest {

    public enum TestEnum implements Marshallable, ReadResolvable<TestEnum> {
        INSTANCE;

        @Override
        public void readMarshallable(@NotNull WireIn wire) throws IORuntimeException {
        }

        @Override
        public void writeMarshallable(@NotNull WireOut wire) {
        }

        @Override
        public EnumTest.TestEnum readResolve() {
            return INSTANCE;
        }
    }

    @Test
    public void testEnum() {
        TextWire wire = new TextWire(Bytes.elasticByteBuffer());
        wire.write("test").object(TestEnum.INSTANCE);
        TextWire wire2 = TextWire.from(wire.toString());
        Object enumObject = wire2.read(() -> "test").object();
        Assert.assertTrue(enumObject == TestEnum.INSTANCE);
    }
}
```

With earlier versions of Wire this test was passing

**Comment by peter-lawrey** [comment]
There is a "feature" of Unsafe.allocateInstance where you can create new
instances of an enum. Reflection by comparison prevents you creating new
instances of an enum. I suspect this worked before due to a bug in the code.
 In this case, for an enum to work you would need to use this feature to
create a new instance of an enum for deserialization but then use
readResolve to use an actual instance. The current version works if you use
a static class instead.(See below)

I would need to understand the use case better.

Do you want to serialize the state of a enum (as this is not Java supports,
I don't have a good example of how this would work)

NOTE: The test uses Yaml rather than JSON serialization as JSON doesn't
support types.

static class TestEnum implements Marshallable, ReadResolvable<TestEnum> {
    static final TestEnum INSTANCE = new TestEnum();

    @Override
    public void readMarshallable(@NotNull WireIn wire) throws
IORuntimeException {
    }

    @Override
    public void writeMarshallable(@NotNull WireOut wire) {
    }

    @Override
    public EnumTest.TestEnum readResolve() {
        return INSTANCE;
    }
}

**Comment by chauhraj** [comment]
It seems like I cannot use Chronicle-Map(CM) and ChronicleQueue(CQ) together. I tried to find version of CQ that uses chronicle-wire v1.3.6 same as that of CM. Is there a workaround other than using different version of CM?

**Comment by peter-lawrey** [comment]
I suggest you use chronicle-bom to find a combination which was released at
the same time.

**Comment by leventov** [comment]
@peter-lawrey 

> There is a "feature" of Unsafe.allocateInstance where you can create new
instances of an enum. Reflection by comparison prevents you creating new
instances of an enum. I suspect this worked before due to a bug in the code.
 In this case, for an enum to work you would need to use this feature to
create a new instance of an enum for deserialization but then use
readResolve to use an actual instance. The current version works if you use
a static class instead.(See below)

To me it doesn't seem related to Unsafe.allocateInstance(). The reason of error seems to me as that before, Chronicle Wire allowed enum classes to implement Marshallable and look more like normal classes in Wire form (`!TestEnum {}`). Now it allows *only* (`!TestEnum.INSTANCE`), and looking at `!TestEnum {}`, it tries to interpret `{}` as enum constant, and fails.

It is not a bug in the former code, it's behaviour change (making behaviour more strict).

If now Wire disallows `!TestEnum {}`, why it serializes enum as such? In the failing test that I given, I don't use "former" format, I just serialize an object and not able to deserialize it back. If now Wire is able to parse only `!TestEnum.INSTANCE`, it should serialize `TestEnum.INSTANCE` as `!TestEnum.INSTANCE`, not `!TestEnum {}` as it does now.

**Comment by peter-lawrey** [comment]
The only reason to use readResolve is to replace one instance with another.
For that other instance to exist, you would have to create a new instance
of an enum which isn't desirable.

It is possible to provide backward support for reading, though this would
need to be thought through so there is not further unintended consequences.

**Comment by chauhraj** [comment]
@peter-lawrey I tried to find chronicle-bom:v1.11.16 for Chronicle Queue but I couldn't find one. I found that CQ v4.0.9 -> chronicle-bom v1.11.15 -> CW v1.3.5
             v4.1.0 -> v1.11.18 -> v1.3.7

I couldn't find CQ which uses v1.11.16 which has CW v1.3.6. So my closest bet would be to use CQ v4.1.0 and force my build to use v1.3.6. Do you guys see anything breaking with this?

**Comment by leventov** [comment]
@chauhraj I've just released [Chronicle Map 3.9.0](https://github.com/OpenHFT/Chronicle-Map/releases/tag/chronicle-map-3.9.0) which should work with latest Chronicle Wire versions.

**Comment by leventov** [comment]
@peter-lawrey OK, I've managed to preserve compatibility by replacing enums with ordinary classes with static final `INSTANCE` constants, and released a new Chronicle Map version which depends on the latest Wire. So, Wire compatibility is not an issue for Chronicle Map anymore (however it might be for other users).

But, I see no justification that the test that I suggested in this topic fails. IMO it is a bug in Wire that this test fails.

**Comment by peter-lawrey** [comment]
I don't think there should be an expectation that new instances of an enum
should ever be created,
This doesn't mean it shouldn't be backwardly compatible, however I consider
it a bug that this was ever allowed.

**Comment by peter-lawrey** [comment]
I believe this should fix backward compatibility.

https://github.com/OpenHFT/Chronicle-Wire/commit/380a6e1e4ca3a0d5e717423c384906fae84189d3

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.20.101](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.20.101), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)
